csrc/host_ir/host_ir.h

```diff
@@ -0,0 +1,118 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <fusion.h>
+#include <ir/base_nodes.h>
+#include <ir/builder.h>
+
+namespace nvfuser {
+
+namespace hir {
+
+/*
+  Host Irs are used to represent a host program. They need to be registered in a
+  HostIrContainer. Each Ir represents a Host data or instruction.
+*/
+
+/*
+  HostUnit represents a Fusion in the Host Program. In other words, it
+  represents a compute graph (or a segment of a larger compute graph)
+  represented by a Fusion that should be compiled and executed as a bulked item
+  from the host perspective.
+
+  This IR can be thought as a thin layer around the class `Fusion`, which
+  furthermore inherits from `Expr` so that it is an "IR" in nvFuser IR
+  semantics.
+
+  This IRs fundamentally allows nested IR structures. It could potentially be
+  useful in other instances than HostIrs.
+
+  Its implementation is minimal, the only specifity being the moethod
+  `fusion_to_execute()` that returns the fusion that the IR represents.
+
+  Note: HostUnit has no I/O itself -- however the Fusion it embbeds has I/O of
+  course, which are not registered in the surrounding HostIrContainer.
+
+  Note: Whether HostUnit should inherit from Expr or Val is debatable. Both are
+  possible, I define it as an Expr for now here but am open to change it.
+*/
+class NVF_API HostUnit : public Expr {
```

Comment:
>  But I'd be aware of YAGNI. I doubt these features will be needed for the foreseeable future. Maybe after fusion definition supports function calls, we'll start to worry about this.

I'm applying this principle the best I can, not to add any unnecessary complexity and to make the PR as simple (as I am able to)

> I understood your intention of supporting compilation-during-execution and arbitrarily nested IR etc. 

In this PR there is no arbitrarily nesting nor async compilation (the compilation is bulked with launching the kernel for now, and only occurs in the executor, there is no IR representing compilation)

> I **think** there's an easier route of first compiling all segments and then using host IR to directly launch compiled kernels that can be identified by kernel names.

I fail to see how it would simplify the current PR